File: lc_plusplus/Search in Rotated Sorted Array.cc
// Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.
//
// (i.e., 0 1 2 4 5 6 7 might become 4 5 6 7 0 1 2).
//
// You are given a target value to search. If found in the array return its index, otherwise return -1.
//
// You may assume no duplicate exists in the array.
//


class Solution {
public:
    int search(vector<int>& nums, int target) {
        for (int i = 0, j = static_cast<int>(nums.size()) - 1; i <= j;) {
            auto m = (i + j) / 2;
            if (target == nums[m]) return m;
            if (nums[m] >= nums[i]) {  // must >= not just > for the case of m == i
                if (target > nums[m] || target < nums[i]) i = m + 1;
                else j = m - 1;
            } else {
                if (target < nums[m] || target > nums[j]) j = m - 1;
                else i = m + 1;
            }
        }
        return -1;
    }
};

File: lc_java/d/Search in Rotated Sorted Array.java
/**
 * Suppose a sorted array is rotated at some pivot unknown to you beforehand.
 * <p>
 * (i.e., 0 1 2 4 5 6 7 might become 4 5 6 7 0 1 2).
 * <p>
 * You are given a target value to search. If found in the array return its index, otherwise return -1.
 * <p>
 * You may assume no duplicate exists in the array.
 */

public class Solution {
  public int search(int[] nums, int target) {
    int lo = 0;
    int hi = nums.length - 1;
    while (lo <= hi) {
      int mid = lo + (hi - lo) / 2;
      if (nums[mid] == target) return mid;
      else if (nums[mid] > target) {
        if (nums[mid] < nums[hi]) hi = mid - 1;
        else {
          if (target >= nums[lo]) hi = mid - 1;
          else lo = mid + 1;
        }
      } else {
        if (nums[mid] > nums[lo]) lo = mid + 1;
        else {
          if (nums[hi] >= target) lo = mid + 1;
          else hi = mid - 1;
        }
      }
    }
    return -1;
  }
}
